src/main/java/com/uber/cadence/internal/replay/ReplayDecisionTaskHandler.java

```diff
@@ -96,71 +103,113 @@ public ReplayDecisionTaskHandler(
 
   private Result handleDecisionTaskImpl(DecisionTaskWithHistoryIterator decisionTaskIterator)
       throws Throwable {
-    HistoryHelper historyHelper = new HistoryHelper(decisionTaskIterator);
-    PollForDecisionTaskResponse decisionTask = historyHelper.getDecisionTask();
-    ReplayDecider decider = createDecider(decisionTask);
-
-    if (decisionTask.isSetQuery()) {
-      RespondQueryTaskCompletedRequest queryCompletedRequest =
-          new RespondQueryTaskCompletedRequest();
-      queryCompletedRequest.setTaskToken(decisionTask.getTaskToken());
-      try {
-        byte[] queryResult = decider.query(historyHelper, decisionTask.getQuery());
-        queryCompletedRequest.setQueryResult(queryResult);
-        queryCompletedRequest.setCompletedType(QueryTaskCompletedType.COMPLETED);
-      } catch (Throwable e) {
-        // TODO: Appropriate exception serialization.
-        StringWriter sw = new StringWriter();
-        PrintWriter pw = new PrintWriter(sw);
-        e.printStackTrace(pw);
-        queryCompletedRequest.setErrorMessage(sw.toString());
-        queryCompletedRequest.setCompletedType(QueryTaskCompletedType.FAILED);
+    PollForDecisionTaskResponse decisionTask = decisionTaskIterator.getDecisionTask();
+
+    Decider decider =
+        stickyTaskListName == null
+            ? createDecider(decisionTask)
+            : cache.getOrCreate(decisionTask, this::createDecider);
+    try {
+      if (decisionTask.isSetQuery()) {
+        return processQuery(decisionTaskIterator, (ReplayDecider) decider);
```

Comment:
I'm confused. My understanding was that Decider is an interface that all its dependencies should use. Downcasting it breaks this model.

Reply:
added a comment to refactor this in a separate CR